engine/src/lib.rs

```diff
@@ -598,6 +603,87 @@ mod contract {
         }
     }
 
+    /// Sets relayer key manager.
+    #[no_mangle]
+    pub extern "C" fn set_key_manager() {
+        let mut io = Runtime;
+        let mut state = state::get_state(&io).sdk_unwrap();
+
+        require_owner_only(&state, &io.predecessor_account_id());
+
+        let key_manager =
+            serde_json::from_slice::<RelayerKeyManagerArgs>(&io.read_input().to_vec())
+                .map(|args| args.key_manager)
+                .sdk_expect(errors::ERR_JSON_DESERIALIZE);
+
+        if state.key_manager == key_manager {
+            sdk::panic_utf8(errors::ERR_SAME_KEY_MANAGER)
+        } else {
+            state.key_manager = key_manager;
+            state::set_state(&mut io, &state).sdk_unwrap();
+        }
+    }
+
+    /// Adds a relayer function call key.
+    #[no_mangle]
+    pub extern "C" fn add_relayer_key() {
+        let mut io = Runtime;
+        let state = state::get_state(&io).sdk_unwrap();
+        require_key_manager_only(&state, &io.predecessor_account_id());
+
+        let public_key = serde_json::from_slice::<RelayerKeyArgs>(&io.read_input().to_vec())
+            .map(|args| args.public_key)
+            .sdk_expect(errors::ERR_JSON_DESERIALIZE);
+        let allowance = Yocto::new(io.attached_deposit());
+        sdk::log!("attached key allowance: {allowance}");
+
+        if allowance.as_u128() < 100 {
+            // TODO: Clarify the minimum amount if check is needed then change error type
+            sdk::panic_utf8(errors::ERR_NOT_ALLOWED);
+        }
+
+        engine::add_function_call_key(&mut io, &public_key);
+
+        let action = PromiseAction::AddFunctionCallKey {
+            public_key,
+            allowance,
+            nonce: 0, // not actually used - depends on block height
+            receiver_id: io.current_account_id(),
+            function_names: "call,submit,submit_with_args".to_string(),
+        };
+        let promise = PromiseBatchAction {
+            target_account_id: io.current_account_id(),
+            actions: vec![action],
+        };
+
+        let promise_id = unsafe { io.promise_create_batch(&promise) };
```

Comment:
Because access keys are stored as `account_id,public_key` in a trie state. As I understand we can add the public there by sending promises only as we don't have direct access to the trie state from the aurora contract.